File: lua/LineArt.lua
function getFontTable()
    oldInput = io.input()
    io.input("font.txt")
    -- First six characters
    height = io.read("*number")
    io.read()
    local map = {}
    local letter = io.read(1)
    io.read()
    while letter ~= nil do
        map[letter] = ""
        -- Read large letter and put in map
        for word = 1, height do
            newPart = io.read()
            map[letter] = map[letter] .. newPart .. "\n"
        end
        -- Get letter
        letter = io.read(1)
        io.read()
    end
    io.input(oldInput)
    return map
end

function printAsciiString(string, fontMap)
    local characterTable = {}
    for word in string:gmatch(".") do
        characterTable[#characterTable+1] = fontMap[word]
    end
    printTableLines(characterTable)
end

function printTableLines(characterTable)
    -- Base case
    ContainsLines = false
    for i, letter in ipairs(characterTable) do
        if(string.find(characterTable[1], "\n") ~= nil) then
            ContainsLines = true
        end
    end
    if ContainsLines == false then
        return
    end
    -- Recursive Case
    local thisLine = ""
    local newCharacterTable = {}
    for i, letter in ipairs(characterTable) do
        -- Find cutpoint between lines
        local wordCutoff = string.find(letter, "\n")
        -- If cutpoint was found
        if wordCutoff ~= nil then
            -- Save next line to print and modify table to not remove next line
            thisLine = thisLine .. string.sub(letter, 0, wordCutoff-1)
            newCharacterTable["size" .. i] = wordCutoff
            newCharacterTable[i] = string.sub(letter, wordCutoff+1, string.len(letter))
        else
            -- Otherwise print just strings
            thisLine = thisLine .. string.rep(" ", characterTable["size" .. i])
        end
    end
    -- Print next line
    print(thisLine)
    printTableLines(newCharacterTable)
end

function lineArtREPL(fontMap)
    while true do
        print("Type in a word to turn into line-art. To exit print-loop press ENTER")
        printWord = io.read()
        if(printWord == "") then
            break
        end
        printAsciiString(printWord, fontMap)
    end
end

function lineArtMaker(fontMap)
    while boundLetter == nil or boundLetter == "" do
        print("Type in a letter to be bound to ART")
        boundLetter = io.read(1)
    end
    -- Flush out rest of io so that the rest of that line isn't included
    io.read()

    print("Binding " .. boundLetter .. " to given art. Create your art here!")
    newArt = ""
    local newArtMap = {}
    local symbolSize = 0
    for i=1,8 do
        newArtMap[i] = io.read()
        if(string.len(newArtMap[i]) > symbolSize) then
            symbolSize = string.len(newArtMap[i])
        end
    end
    for i, word in pairs(newArtMap) do
        print(word .. " : " .. symbolSize - string.len(word))
        newArt = newArt .. word .. string.rep(" ", symbolSize - string.len(word)) .. "\n"
    end
    fontMap[boundLetter] = newArt
end

function saveLineArt(fontMap)
    print("Saving...")
    io.output("font.txt")
    io.write(height .. "\n")
    for i, word in pairs(fontMap) do
        io.write(i .. "\n" .. word)
    end
    print("Done!")
    io.output():close()
end

local fontMap = getFontTable()

print("Welcome to the CPSC4100 Line Art program!")
while true do
    print("Type 1 to enter the line-art loop!")
    print("Type 2 to enter the line-art editor!")
    print("Type 3 to save your current keybind setup and override the existing text file!")
    print("Press enter without input to exit!")
    choice = io.read()
    if(choice == "1") then
        lineArtREPL(fontMap)
        print()
        print("Welcome to the CPSC4100 Line Art program!")
    elseif(choice == "2") then
        lineArtMaker(fontMap)
        print("Your art has been added to the current lineart map! Press 3 in the main menu to save your new keybind!")
        print()
        print("Welcome to the CPSC4100 Line Art program!")
    elseif(choice == "3") then
        saveLineArt(fontMap)
        print()
    elseif(choice == "") then
        break
    end
end

File: php/lineArt.php
<?php

function getFontTable()
{
	//using our unified font document
	$myFile = fopen("font.txt", 'r');
	$height = 8;
	$map = array();
	fgets($myFile);
	$get = fgets($myFile);
	//$letter = fgets($myFile);
	$letter = substr($get,0,-1);
	$map[$letter]="";
	//getting each letter bound into an array key
	if(isset($map[$letter]))
	{
	while($letter != NULL)
		{
			$map[$letter] = "";
			for ($word = 1; $word<=$height;$word++)
			{
				$newPart = fgets($myFile);
				$map[$letter] = $map[$letter].$newPart;
				
			}
			$get= fgets($myFile);
			$letter = substr($get,0,-1);
		}
	}
	//close the file and return the array
	fclose($myFile);
	return $map;
}

function printAsciiString($string,$fontMap)
{
	//im using a new array to simplify things a little
	$characterTable=[];
	//this array is indexed with numbers
	$sepString=str_split($string);
	for($i =0; $i<strlen($string);$i++)
	{
		$characterTable[$i]=$fontMap[$sepString[$i]];

	}
	checkTableLines($characterTable);
}

//getting to output on one line
function checkTableLines($characterTable)
{
	//base case
	$containsLines = False;
	if(strpos($characterTable[0],"\n") !==False)
	{
		$containsLines = true;
	}
	if($containsLines==false)
	{
		return;
	}
	//recursive case
		$thisLine="";
		//printing a line at a certain height
		$newCharacterTable=[];
		//this new table will contain the previous table with the previous line cut
		for ($i=0;$i<sizeof($characterTable);$i++)	
		{
			//finding the end point of each line
			$wordEnd = strpos($characterTable[$i], "\n");
			if ($wordEnd != NULL)
			{
				$thisLine = $thisLine.substr($characterTable[$i],0,$wordEnd);
				$newCharacterTable[$i]=substr($characterTable[$i],$wordEnd+1,strlen($characterTable[$i]));
			}
			else
			{
				$thisLine=$thisLine.str_repeat(" ", $key=array_search($characterTable[$i],$characterTable));
			}
				
		}
		//print an individual line each loop
		echo $thisLine,"\n";
		//call again with new array 
		checkTableLines($newCharacterTable);
	
}

//allowing for continuous input
function lineArtREPL($fontMap)
{
	while (true)
	{
		print("type in a word to turn it into line-art. to exit print-loop press ENTER\n");
		$printword = readline();
		if($printword=="")
			break;
		else
			printAsciiString($printword, $fontMap);
	}
}

$fontMap = getFontTable();
echo "welcome to the CPSC4100 Line Art Program!","\n";
while(true)
{
	echo "\n","Type 1 to enter the line-art loop!","\n";
    echo "Type 2 to enter the line-art editor!","\n";
    echo "Type 3 to save your current keybind setup and override the existing text file!","\n";
    echo "Press enter without input to exit!","\n";
	$user=readline();
	if($user=="1")
	{
		lineArtREPL($fontMap);
		echo "welcome to the CPSC4100 line art program!";
	}
	elseif($user=="2")
	{
		lineArtMaker($fontMap);
		echo "your art has been aded to the current lineart map. press 3 in the main menu to save your keybinding","\n";
		echo "welcome to the cpsc4100 line art program!","\n";	
	}
	elseif($user=="3")
	{
		saveLineArt($fontMap);
		echo "\n";
	}
	elseif($user=="")
	{
		break;
	}
	
}
?>
